### PR TITLE
Refactoring JDispatcher and JEvent, deprecating JObserver and JObservable

### PR DESCRIPTION
**Pull request by Hackwar**
This pull request refactors JDispatcher and JEvent to directly extend JObject and deprecates JObservable and JObserver.

JObservable and JObserver are both broken and should not be used anymore.

Besides refactoring the code, JDispatcher and JEvent were both covered by Unittests, which should have full coder coverage.

The discussion regarding this change can be found here: https://groups.google.com/group/joomla-dev-platform/browse_thread/thread/4b195ce76d200e3e#

### CODE REVIEW

**Comment by eddieajau** [comment]
Thanks Hannes. Just on the code style side of things we need a new line at the end of the files now (there's a new sniff for that), and also for the new code that is going into JDispatcher and others, can you make sure the protected properties don't have the underscore.

**Comment by eddieajau** [comment]
Hannes, I just noticed JEditor and JAuthentication also extend JObservable.

**Comment by Hackwar** [comment]
Ok, will change the new line at the end. Regarding the underscore: Since the variables are not new and might be used somewhere, I didn't want to change them. Remember that its basically just removing the inheritance and copying the code into the child-class. But if you want, I can change it anyway.

I will also take a sweep and look at all classes that might extend JObservable.

**Comment by Hackwar** [comment]
Refactored JEditor and JAuthentication. I didn't update my sniffs yet, but the files should all have an empty line. Can you decide on renaming the properties?

**Comment by eddieajau** [comment]
Doesn't merge.  Please leave the tests in until the classes are removed. Thanks.

**Comment by Hackwar** [comment]
Added the tests back in and resolved the conflicts. This should merge now. Could you try again?

**Comment by elkuku** [comment]
**Checkstyle error details:**
`libraries/joomla/base/observable.php:197`
Please end your files with an empty line.
`libraries/joomla/event/event.php:34`
Constructor and destructor comments must not have a @return tag
`libraries/joomla/html/editor.php:46`
Whitespace found at end of line
`libraries/joomla/user/authentication.php:491`
Please end your files with an empty line.

File: libraries/joomla/event/dispatcher.php [hunk 1]
```diff
@@ -77,7 +111,7 @@ class JDispatcher extends JObservable
 		}
 		else
 		{
-			JError::raiseWarning('SOME_ERROR_CODE', JText::sprintf('JLIB_EVENT_ERROR_DISPATCHER', $handler));
+			throw new Exception(JText::sprintf('JLIB_EVENT_ERROR_DISPATCHER', $handler));
```
Comment (eddieajau) [inline libraries/joomla/event/dispatcher.php]: I think we need to keep the JError here until we remove it next year.  The CMS will still use this version of the platform.

Comment (Hackwar) [inline libraries/joomla/event/dispatcher.php]: Done

**Comment by eddieajau** [comment]
Dang it, so close!  Hannes, the offending check-style is:

FILE: ...itories/github/joomla-platform/libraries/joomla/user/authentication.php
FOUND 1 ERROR(S) AFFECTING 1 LINE(S)
 490 | ERROR | Please end your files with an empty line.

Could you fix that and we'll get this fella knocked on the head once and for all.  Thanks.

**Comment by Hackwar** [comment]
done

**Comment by eddieajau** [comment]
Thanks

**Pull request merged**